src/rpc/rpc_impl/query.rs

```diff
@@ -21,15 +31,256 @@ impl<S: Store> MercuryRpcImpl<S> {
+                cell.cell_output.lock().code_hash() == config.script.code_hash()
+                    && cell.cell_output.lock().hash_type() == config.script.hash_type()
+                    && cell.cell_output.type_().is_none()
+            })
+            .map(|cell| {
+                let capacity: u64 = cell.cell_output.capacity().unpack();
+                capacity
+            })
+            .sum();
+        Ok(spendable_ckb_balance)
+    }
+
+    pub(crate) fn acp_spendable_udt_balance(&self, addr: &Address, udt_hash: H256) -> Result<u128> {
+        let cells = self.get_sp_detailed_cells(addr)?;
+        let config = self.config.get(special_cells::ACP).unwrap();
+        let spendable_udt_balance = cells
+            .0
+            .iter()
+            .filter(|cell| {
+                cell.cell_output.lock().code_hash() == config.script.code_hash()
+                    && cell.cell_output.lock().hash_type() == config.script.hash_type()
+                    && cell.cell_output.type_().is_some()
+            })
+            .filter(|cell| {
+                let type_script = cell.cell_output.type_().to_opt().unwrap();
+                let type_script_hash: [u8; 32] = type_script.calc_script_hash().unpack();
+                type_script_hash == udt_hash.0
+            })
+            .map(|cell| utils::decode_udt_amount(&cell.cell_data.raw_data()))
+            .sum();
+        Ok(spendable_udt_balance)
+    }
+
+    pub(crate) fn cheque_spendable_udt_balance(
+        &self,
+        addr: &Address,
+        udt_hash: H256,
+    ) -> Result<u128> {
+        let script = address_to_script(addr.payload());
+        let pubkey_hash = H160::from_slice(&script.args().raw_data()[0..20]).unwrap();
+        let cells = self.get_sp_detailed_cells(addr)?;
+        let config = self.config.get(special_cells::CHEQUE).unwrap();
+        let spendable_udt_balance = cells
+            .0
+            .iter()
+            .filter(|cell| {
+                cell.cell_output.lock().code_hash() == config.script.code_hash()
+                    && cell.cell_output.lock().hash_type() == config.script.hash_type()
+                    && cell.cell_output.type_().is_some()
+            })
+            .filter(|cell| {
+                let type_script = cell.cell_output.type_().to_opt().unwrap();
+                let type_script_hash: [u8; 32] = type_script.calc_script_hash().unpack();
+                type_script_hash == udt_hash.0
+            })
+            .filter(|cell| {
+                // filter sender pubkey_hash
+                let lock_args = cell.cell_output.lock().args().raw_data();
+                lock_args.len() == 40 && lock_args[20..40] == pubkey_hash.0
+            })
+            .filter(|cell| {
+                let cell_epoch = RationalU256::from_u256(cell.epoch_number.clone());
+                let cheque_since = RationalU256::from_u256(self._cheque_since.clone());
+                let current_epoch = &*CURRENT_EPOCH.read();
```

Comment:
Do not use RwLock like this. You should get the current epoch at the beginning of the function as I suggested ahead.

Reply:
updated.